iroh-resolver/src/balanced_tree.rs

```diff
@@ -0,0 +1,388 @@
+
+        // "manually" build expected stream output
+        let mut expect = Vec::new();
+        let chunk = chunk_stream.next().await.unwrap().unwrap().freeze();
+        let node = TreeNode::Leaf(chunk.clone());
+        let (cid, bytes, _) = node.encode().unwrap();
+        expect.push((cid, bytes.clone()));
+        let got = stream_balanced_tree(test_chunk_stream(1), 3);
+        tokio::pin!(got);
+        ensure_equal(expect, got).await;
+    }
+
+    #[tokio::test]
+    async fn balanced_tree_test_height_one() {
+        let num_chunks = 3;
+        let degrees = 3;
+        let chunk_stream = test_chunk_stream(num_chunks);
+        tokio::pin!(chunk_stream);
+
+        // "manually" build expected stream output
+        let mut expect = Vec::new();
+        let mut links = Vec::new();
+        while let Some(chunk) = chunk_stream.next().await {
+            let chunk = chunk.unwrap().freeze();
+            let node = TreeNode::Leaf(chunk);
+            let (cid, bytes, len) = node.encode().unwrap();
+            links.push((cid, len));
+            expect.push((cid, bytes));
+        }
+        // let root = TreeNode::Root(links);
+        let root = TreeNode::Stem(links);
+        let (cid, bytes, _) = root.encode().unwrap();
+        expect.push((cid, bytes));
+
+        let got = stream_balanced_tree(test_chunk_stream(num_chunks), degrees);
+        tokio::pin!(got);
+        ensure_equal(expect, got).await;
+    }
+
+    #[tokio::test]
+    async fn balanced_tree_test_height_two_full() {
+        let degrees = 3;
+        let num_chunks = 9;
+        let chunk_stream = test_chunk_stream(num_chunks);
+        tokio::pin!(chunk_stream);
+
+        // "manually" build expected stream output
+        let mut expect = Vec::new();
+        let mut links_one = Vec::new();
+        let mut links_two = Vec::new();
+        while let Some(chunk) = chunk_stream.next().await {
+            let chunk = chunk.unwrap().freeze();
+            let node = TreeNode::Leaf(chunk);
+            let (cid, bytes, len) = node.encode().unwrap();
+            links_one.push((cid, len));
+            expect.push((cid, bytes));
+            if links_one.len() == degrees {
+                let links = std::mem::take(&mut links_one);
+                let stem = TreeNode::Stem(links);
+                let (cid, bytes, len) = stem.encode().unwrap();
+                links_two.push((cid, len));
+                expect.push((cid, bytes));
+            }
+        }
+        // let root = TreeNode::Root(links_two);
+        let root = TreeNode::Stem(links_two);
+        let (cid, bytes, _) = root.encode().unwrap();
+        expect.push((cid, bytes));
+
+        let got = stream_balanced_tree(test_chunk_stream(num_chunks), degrees);
+        tokio::pin!(got);
+        ensure_equal(expect, got).await;
+    }
+
+    #[tokio::test]
+    async fn balanced_tree_test_height_two_not_full() {
+        let num_chunks = 7;
+        let degrees = 3;
```

Comment:
I would suggest to make the tests such that we can run over multiple different degrees, and num chunks in some generic way, so we can increase the coverage here

Reply:
I don't see any test with other degrees than 3?

Reply:
@ramfox ping

Reply:
yea, thank you for re-pinging I somehow missed this during both passes...